From 62364851437277c265e8300543e0d8852cc0e292 Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Sat, 11 Sep 2021 17:54:32 +0200
Subject: [PATCH 712/985] evl/poll: fix off-by-one in attaching the watchpoints

This bug would cause the last watchpoint from the poll table to be
left unexpectedly detached, which in turn would break clear_wait().

At this chance, clarify some naming in evl_drop_poll_table().

Signed-off-by: Philippe Gerum <rpm@xenomai.org>
---
 kernel/evl/poll.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/kernel/evl/poll.c b/kernel/evl/poll.c
index abdb37d09..b3255ac8f 100644
--- a/kernel/evl/poll.c
+++ b/kernel/evl/poll.c
@@ -121,11 +121,11 @@ void __evl_signal_poll_events(struct evl_poll_head *head,
 }
 EXPORT_SYMBOL_GPL(__evl_signal_poll_events);
 
-void evl_drop_poll_table(struct evl_thread *thread)
+void evl_drop_poll_table(struct evl_thread *curr)
 {
 	struct evl_poll_watchpoint *table;
 
-	table = thread->poll_context.table;
+	table = curr->poll_context.table;
 	if (table)
 		evl_free(table);
 }
@@ -485,8 +485,10 @@ static int collect_events(struct poll_group *group,
 			if (ret)
 				return -EFAULT;
 			u_set++;
-			if (++count >= maxevents)
+			if (++count > maxevents) {
+				count = maxevents;
 				break;
+			}
 		}
 	}
 
-- 
2.30.2

